#35 feat : 이메일 인증 기능 추가

uri010 opened the pull request:
## 요약
- build.gradle에 redis와 mail 의존성 추가
- 이메일 인증 관련된 로직을 담은 MailService 추가
- 이메일 인증 관련 설정을 담은 MailConstant 추가
- MailService의 로직을 호출할 MailController 추가
- 인증 코드를 관리할 RedisService 추가
- Redis에서 key값에 해당하는 값이 저장되어 있지 않는 경우 NoMatchedKeyException 예외 처리 
- 이메일 인증 관련 로직 수행 중 발생하는 MailSendException과 MessagingException을 ExceptionHandler에서 처리 
- UserConstant에서 LoginConstant로 이름 변경

## 변경 내용
### 1. build.gradle
```
implementation 'org.springframework.boot:spring-boot-starter-mail'
implementation 'org.springframework.boot:spring-boot-starter-data-redis'
```
- mail 관련 기능 사용을 위해 의존성 추가
- 이메일 인증 코드를 인메모리 데이터 저장소인 redis에 저장하기 위해 의존성 추가
    - redis는 데이터를 key-value 형태로 저장해 쿼리를 사용할 필요도 없고, 메모리에서 데이터를 처리하기 때문에 속도가 빠른 장점이 있음
    - 이메일 인증 코드의 경우 제한 시간이 짧고 서버가 꺼지면 무의미해지는 값이기 때문에 redis 사용

### 2. MailConstant
```java
@ConfigurationProperties(prefix = "mail.config")
@ConstructorBinding
public class MailConstant {

	public static long EXPIRE_TIME;
	public static int AUTH_CODE_LENGTH;

	public MailConstant(long expireTime, int authCodeLength) {
		EXPIRE_TIME = expireTime;
		AUTH_CODE_LENGTH = authCodeLength;
	}
}
```
- 이메일 인증 관련 설정 정보를 application.properties에서 가져오는 클래스
- `EXPIRE_TIME` : 인증 코드 유효 시간을 의미
- `AUTH_CODE_LENGTH` : 인증 코드 길이를 의미

### 3. MailController
```java
@RestController
@RequestMapping("/email")
@RequiredArgsConstructor
public class MailController {
	private final MailService mailService;

	@PostMapping
	public void sendAuthEmail(@RequestParam @Email String email) throws MessagingException {
		mailService.sendAuthEmail(email);
	}

	@PostMapping("/auth")
	public void sendEmailAuthCode(@RequestParam @Email String email, @RequestParam String authCode) {
		mailService.checkEmailAuthCode(email, authCode);
	}
}
```
- 이메일 인증 부분은 MailController에서 처리

### 4. MailService
```java
@Service
@RequiredArgsConstructor
public class MailService {
	private final JavaMailSender mailSender;
	private final RedisService redisService;
	private final MailConstant mailConstant;

	public void sendAuthEmail(String email) throws MessagingException {
		String authCode = createAuthCode(mailConstant.AUTH_CODE_LENGTH);
		log.info("인증 코드 길이 ={}", mailConstant.AUTH_CODE_LENGTH);
		log.info("인증 코드={}", authCode);

		MimeMessage mail = mailSender.createMimeMessage();
		String mailContent = "<h1>[이메일 인증]</h1>"
			+ "<br>"
			+ "<h3>이메일 인증 번호 : " + authCode + "</h3>";

		mail.setSubject("회원가입 이메일 인증 ", "utf-8");
		mail.setText(mailContent, "utf-8", "html");
		mail.addRecipient(Message.RecipientType.TO, new InternetAddress(email));
		log.info("인증 이메일 설정 완료");
		try {
			mailSender.send(mail);
		}
		catch (Exception e) {
			e.printStackTrace();
		}
		redisService.setDataExpire(email, authCode, mailConstant.EXPIRE_TIME);
	}

	public void checkEmailAuthCode(String email, String authCode) {
		if (!redisService.checkData(email, authCode)) {
			throw new NoMatchedKeyException(ErrorCode.NO_MATCHING_AUTH_CODE);
		}
	}

	private String createAuthCode(int size) {
		Random random = new Random();
		StringBuffer buffer = new StringBuffer();
		int num = 0;

		while (buffer.length() < size) {
			num = random.nextInt(10);
			buffer.append(num);
		}

		return buffer.toString();
	}

}
```
- `sendAuthEmail()` : 이메일을 파라미터로 받아 해당 이메일로 인증 코드가 담긴 메일을 보내는 메소드
    - JavaMailSender 설정 관련된 예외인 MessagingException은 checkedException으로 메소드에서 던져줌
- `checkEmailAuthCode()` : RedisService의 checkData 메소드를 통해 파라미터로 받은 email에 해당하는 인증 코드가 redis에 저장되어 있는지 확인
    - email에 해당하는 인증 코드가 없으면 NoMatchedException 예외를 던짐
- `createAuthCode()` : 파라미터로 받은 길이만큼의 랜덤한 숫자를 생성해서 String으로 반환

### 5. RedisService
```java
@Service
@RequiredArgsConstructor
public class RedisService {
	private final StringRedisTemplate stringRedisTemplate;

	public String getData(String key) {
		ValueOperations<String, String> valueOperations = stringRedisTemplate.opsForValue();
		return valueOperations.get(key);
	}

	public void setDataExpire(String key, String value, long duration) {
		ValueOperations<String, String> valueOperations = stringRedisTemplate.opsForValue();
		Duration expireDuration = Duration.ofSeconds(duration);
		valueOperations.set(key, value, expireDuration);
	}

	public boolean checkData(String key, String value) {
		String data = getData(key);

		if (data != null && data.equals(value)) {
			deleteData(key);
			return true;
		}
		return false;
	}

	public void deleteData(String key) {
		stringRedisTemplate.delete(key);
	}
}
```
- Redis 데이터 처리와 관련된 로직을 모아놓은 클래스
- `getData()` : key값에 해당하는 데이터를 반환, 없으면 null
- `setDataExpire()` : key와 value 값을 일정 시간 동안만 저장
- `checkData()` : 파라미터로 전달된 key에 매핑된 data를 가져오고 그 data와 value가 일치한지 확인, 일치하면 redis에서 해당 key-value를 삭제하고 true 반환
- `deleteData()` : redis에서 key값에 해당하는 값을 삭제

### 6. GlobalExceptionHandler
```java
@ExceptionHandler(MailSendException.class)
	public ResponseEntity handleMailSendException(MailSendException ex) {
		return ResponseEntity
			.status(HttpStatus.NOT_IMPLEMENTED)
			.body(new ErrorResponse(HttpStatus.NOT_IMPLEMENTED, "인증 메일 전송에 실패했습니다."));
	}
	@ExceptionHandler(MessagingException.class)
	public ResponseEntity handleMessagingException(MessagingException ex) {
		return ResponseEntity
			.status(HttpStatus.NOT_IMPLEMENTED)
			.body(new ErrorResponse(HttpStatus.NOT_IMPLEMENTED, "인증 메일 설정값이 잘못되었습니다."));
	}

	@ExceptionHandler(NoMatchedKeyException.class)
	public ResponseEntity handleNoMatchedKeyException(NoMatchedKeyException ex) {
		return ResponseEntity
			.status(ex.getErrorCode().getStatus())
			.body(new ErrorResponse(ex.getErrorCode().getStatus(), ex.getErrorCode().getMessage()));
	}
```
- JavaMailSender의 send 메소드의 MailSendException 예외 처리 
- JavaMailSender 설정과 관련된 MessagingException 예외 처리
- Redis에 일치하는 값이 없는 경우 NoMatchedKeyException 예외 처리 

### 7. NoMatchedKeyException
```java
@Getter
@RequiredArgsConstructor
public class NoMatchedKeyException extends RuntimeException {
	private final ErrorCode errorCode;
}
```
### 8. ErrorCode
```java
  NO_MATCHING_AUTH_CODE(HttpStatus.BAD_REQUEST, "인증코드가 일치하지 않습니다."),
```
- 인증코드가 일치하지 않은 경우의 예외 코드

### 9. SharedCalenderApplication
```java
@EnableConfigurationProperties({LoginConstant.class, MailConstant.class})
@SpringBootApplication
public class SharedCalendarApplication {
```
- 이메일 인증 관련 설정 정보를 properties에서 가져오기 위해 EnableConfigurationProperties에 MailConstant 클래스 추가 

### 10. LoginConstant
- 유저보다는 로그인과 관련된 설정 정보라고 판단해 UserConstant에서 LoginConstant로 이름 수정

## Issue number/URL
#35 Feat : 이메일 인증

uri010 committed:
#35 feat : 이메일 인증 기능 추가

build.gradle에 redis와 mail 의존성 추가
이메일 인증 관련된 로직을 담은 MailService 추가
이메일 인증 관련 설정을 담은 MailConstant 추가
MailService의 로직을 호출할 MailController 추가
인증 코드를 관리할 RedisService 추가
Redis에서 key값에 해당하는 값이 저장되어 있지 않는 경우 NoMatchedKeyException 예외 처리
이메일 인증 관련 로직 수행 중 발생하는 MailSendException과 MessagingException을 ExceptionHandler에서 처리
UserConstant에서 LoginConstant로 이름 변경

uri010 committed:
#35 fix : 불필요한 코드 삭제 및 HttpStatus를 static import 처리

uri010 commented:
## 변경된 내용
- MailService의 sendAuthEmail 메소드에서 mailSender.send(mail);을 에러 메시지 확인을 위해  try-catch로 묶은 것을 삭제
- HttpStatus를 static import 처리

yunah7 commented on src/main/java/study/sharedcalendar/service/MailService.java:
```diff
@@ -0,0 +1,64 @@
+		}
+	}
+
+	private String createAuthCode(int size) {
```
다른 이슈에서 RandomStringUtils 관련 gradle을 추가해두어서 randomNumeric 메소드를 사용해도 될 것 같습니다!

sunwootest commented:
pr 첫번째 설명하는 코멘트 마크다운 java로 변경했습니다 :) ` 3번치고 뒤에 java 치면 java 스타일로 코드 읽을 수 있어요!

sunwootest commented on src/main/java/study/sharedcalendar/controller/MailController.java:
```diff
@@ -0,0 +1,29 @@
+import lombok.RequiredArgsConstructor;
+import study.sharedcalendar.service.MailService;
+
+@RestController
+@RequestMapping("/email")
+@RequiredArgsConstructor
+public class MailController {
+	private final MailService mailService;
+
+	@PostMapping
+	public void sendAuthEmail(@RequestParam @Email String email) throws MessagingException {
+		mailService.sendAuthEmail(email);
+	}
+
+	@PostMapping("/auth")
+	public void sendEmailAuthCode(@RequestParam @Email String email, @RequestParam String authCode) {
+		mailService.checkEmailAuthCode(email, authCode);
+	}
+}
```
제 생각에는 이 내용은 userController로 들어가도 될 것 같네요 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/MailService.java:
```diff
@@ -0,0 +1,64 @@
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class MailService {
+	private final JavaMailSender mailSender;
+	private final RedisService redisService;
+	private final MailConstant mailConstant;
+
+	public void sendAuthEmail(String email) throws MessagingException {
+		String authCode = createAuthCode(mailConstant.AUTH_CODE_LENGTH);
+		log.info("인증 코드 길이 ={}", mailConstant.AUTH_CODE_LENGTH);
+		log.info("인증 코드={}", authCode);
+
+		MimeMessage mail = mailSender.createMimeMessage();
+		String mailContent = "<h1>[이메일 인증]</h1>"
+			+ "<br>"
+			+ "<h3>이메일 인증 번호 : " + authCode + "</h3>";
+
+		mail.setSubject("회원가입 이메일 인증 ", "utf-8");
+		mail.setText(mailContent, "utf-8", "html");
+		mail.addRecipient(Message.RecipientType.TO, new InternetAddress(email));
+		log.info("인증 이메일 설정 완료");
+		mailSender.send(mail);
+		redisService.setDataExpire(email, authCode, mailConstant.EXPIRE_TIME);
+	}
```
로깅을 적용하셨네요. 여기서 알아보면 좋을 내용들 몇가지 알려드릴게요 :) 
- sout으로 로그를 남기지 말고 왜 Logger를 사용해야할까?
- 로그에도 수준이 있는데 **스프링 프로파일**을 활용해서 어떤식으로 로그를 이용할 수 있을까?
- 실제 개발환경에서 로그는 언제 남겨야할까?

이 정도 알아보시면 꽤 도움될겁니다!

sunwootest replied:
sendAuthEmail이라는 메소드를  **이메일 전송**과 **인증코드 전송**이라는 관심사로 분리할 수 있을 것 같네요. 이렇게 분리하면 **이메일 전송**이라는 관심사로 만든 메소드를 재활용할 수 있겠죠? 

mailContent와 같은 String 변수도 메소드로 분리할 수 있겠네요 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/RedisService.java:
```diff
@@ -0,0 +1,42 @@
+	private final StringRedisTemplate stringRedisTemplate;
+
+	public String getData(String key) {
+		ValueOperations<String, String> valueOperations = stringRedisTemplate.opsForValue();
+		return valueOperations.get(key);
```
이정도는 한줄로 표현해도 괜찮아보입니다 :) 
```java
return stringRedisTemplate.opsForValue().get(key)
```

sunwootest commented on src/main/java/study/sharedcalendar/service/RedisService.java:
```diff
@@ -0,0 +1,42 @@
+@Service
+@RequiredArgsConstructor
+public class RedisService {
+	private final StringRedisTemplate stringRedisTemplate;
```
RedisTemplate은 StringRedisTemplate으로 사용할 수도 있지만 RedisTemplate\<String\>과 같이 제네릭으로 사용할 수도 있습니다 :) 참고만 하세요!

sunwootest replied:
\@Configuration을 이용해 RedisTemplate에 대한 빈설정을 config 형태로 저장할 수도 있습니다. 이러면 굳이 StringRedisTemplate을 사용하지 않아도 되겠죠? 이 부분은 Redis를 나중에 섬세하게 적용해야할 때 필요할 수 있으니 기억해두세요!

sunwootest reviewed (commented):
이제 새로운 기술들을 적용하는데 두려움이 많이 사라진 것 같아보이네요!

uri010 committed:
Merge branch 'develop' into feature/35

# Conflicts:
#	src/main/java/study/sharedcalendar/constant/ErrorCode.java
#	src/main/java/study/sharedcalendar/constant/LoginConstant.java
#	src/main/java/study/sharedcalendar/controller/UserController.java
#	src/main/java/study/sharedcalendar/dto/SignUpReq.java
#	src/main/java/study/sharedcalendar/service/LoginService.java
#	src/main/java/study/sharedcalendar/service/UserService.java
#	src/main/resources/mybatis-mapper/UserMapper.xml

uri010 committed:
#35 fix : 이메일을 보내는 메소드와 이메일 내용을 만드는 메소드 추가 및 코드 정리

MailController를 삭제하고 UserController로 옮김
WebConfig의 로그인 상태를 확인하지 않을 uri에 email 관련 uri 추가
RedisService 코드 간소화
MailService에 이메일 내용을 반환하는 메소드와 이메일을 보내는 메소드 추가
UserConstant를 LoginConstant로 변경
AutoLoginInterceptor 불필요한 코드 삭제

uri010 commented:
## 요약 
- MailController를 삭제하고 UserController로 옮김
- WebConfig의 로그인 상태를 확인하지 않을 uri에 email 관련 uri 추가
- RedisService 코드 간소화
- MailService에 이메일 내용을 반환하는 메소드와 이메일을 보내는 메소드 추가
- UserConstant를 LoginConstant로 변경
- AutoLoginInterceptor 불필요한 코드 삭제 (private final UserService userService;)

## 변경 내용
### 1. MailService
``` java
      public void sendAuthEmail(String email) throws MessagingException {
		String authCode = createAuthCode(mailConstant.AUTH_CODE_LENGTH);
		log.info("인증 코드 길이 ={}", mailConstant.AUTH_CODE_LENGTH);
		log.info("인증 코드={}", authCode);

		String mailContent = createAuthMailContent(authCode);
		sendEmail("회원가입 이메일 인증", mailContent, email);

		redisService.setDataExpire(email, authCode, mailConstant.EXPIRE_TIME);
	}

	private String createAuthMailContent(String authCode) {
		return "<h1>[이메일 인증]</h1>"
			+ "<br>"
			+ "<h3>이메일 인증 번호 : " + authCode + "</h3>";
	}

	private void sendEmail(String subject, String mailContent, String email) throws MessagingException {
		MimeMessage mail = mailSender.createMimeMessage();
		mail.setSubject(subject, "utf-8");
		mail.setText(mailContent, "utf-8", "html");
		mail.addRecipient(Message.RecipientType.TO, new InternetAddress(email));
		log.info("인증 이메일 설정 완료");
		mailSender.send(mail);
	}
```

>이외의 변경 내용은 간단해서 생략함

sunwootest commented on src/main/java/study/sharedcalendar/service/MailService.java:
```diff
@@ -0,0 +1,72 @@
+	private final RedisService redisService;
+	private final MailConstant mailConstant;
+
+	public void sendAuthEmail(String email) throws MessagingException {
+		String authCode = createAuthCode(mailConstant.AUTH_CODE_LENGTH);
+		log.info("인증 코드 길이 ={}", mailConstant.AUTH_CODE_LENGTH);
+		log.info("인증 코드={}", authCode);
+
+		String mailContent = createAuthMailContent(authCode);
+		sendEmail("회원가입 이메일 인증", mailContent, email);
+
+		redisService.setDataExpire(email, authCode, mailConstant.EXPIRE_TIME);
+	}
```
인증코드 관련 내용을 분리할 수 있을 것 같네요. email 전송만 담당하는 메소드를 만들고 email에 들어갈 내용은 다른 메소드로 분리하는 것이 나아보입니다. 이렇게 작성하면 결국 인증코드 전송용 이메일전송 메소드가 됩니다 :)

uri010 replied:
제목, 메일내용, 보낼 이메일을 파라미터로 받아 이메일을 전송하는 sendEmail()이라는 메소드를 만들어두고 sendAuthEmail에서 사용했습니다. email에 들어갈 내용을 다른 메소드로 분리하라는게 어떤 건지 잘 모르겠습니다🥲

sunwootest replied:
제가 로그로 판단해서 잘못 봤네요. 잘짜셨습니다. 다만 sendEmail 메소드에인증 이메일 설정 완료라는 로그는 다른 것으로 바꾸는 것이 나아보입니다 :)

sunwootest commented on src/main/java/study/sharedcalendar/service/MailService.java:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+
+	private String createAuthCode(int size) {
+		Random random = new Random();
+		StringBuffer buffer = new StringBuffer();
+		int num = 0;
+
+		while (buffer.length() < size) {
+			num = random.nextInt(10);
+			buffer.append(num);
+		}
+
+		return buffer.toString();
+	}
```
희박하지만 Java에서 Random을 사용할 경우 동시성 이슈가 발생할 수 있습니다. 이를 해결하는 다른 난수 생성방법이 존재하는데 이런 것도 알아보시면 좋을 것 같습니다.

sunwootest reviewed (approved):
approve 드립니다 :)

uri010 committed:
#35 fix : 랜덤 인증 코드 생성시 RandomStringUtils 사용, 로그 메시지 변경

동시성 이슈 해결을 위해 랜덤 인증 코드 생성시 RandomStringUtils.randomNumeric() 메소드 사용
sendEmail의 로그 메시지 변경

uri010 commented:
## 요약
- 동시성 이슈 해결을 위해 랜덤 인증 코드 생성시 RandomStringUtils.randomNumeric() 메소드 사용
- sendEmail 메소드의 로그 메시지를 "인증 이메일 설정 완료"에서 "이메일 설정 완료"로 변경
